From 42179711490810567303de62a1ccb680f10d6f5a Mon Sep 17 00:00:00 2001
From: Andrew Jeffery <andrew@aj.id.au>
Date: Tue, 25 Jun 2019 09:51:50 +0930
Subject: [PATCH 4/5] mctp-lpc: Only trigger IRQ on ODR write

Userspace must write ODR in conjunction with STR for the host to
correctly handle the IRQ.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>
---
 drivers/misc/mctp-lpc.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/mctp-lpc.c b/drivers/misc/mctp-lpc.c
index 7346807..2e8ac52 100644
--- a/drivers/misc/mctp-lpc.c
+++ b/drivers/misc/mctp-lpc.c
@@ -255,9 +255,14 @@ static ssize_t mctp_lpc_write(struct file *filp, const char __user *buf,
 
 	regmap_read(priv->map, LPC_STR4, &str);
 	dev_dbg(dev, "Triggering SerIRQ (current str=0x%x)\n", str);
-	/* Trigger Host IRQ */
-	regmap_update_bits(priv->map, LPC_HICRC, LPC_KCS4_IRQ_HOST,
-			   LPC_KCS4_IRQ_HOST);
+
+	/*
+	 * Trigger Host IRQ on ODR write. Do this after any STR write in case
+	 * we need to write ODR to indicate an STR update (which we do).
+	 */
+	if (*ppos == 0)
+		regmap_update_bits(priv->map, LPC_HICRC, LPC_KCS4_IRQ_HOST,
+				   LPC_KCS4_IRQ_HOST);
 
 	return count;
 }
-- 
1.8.3.1

